File: aegis-python/ws.py
from __future__ import annotations

import random
import time
from typing import Callable

from websocket import WebSocketApp


class AegisWebSocketClient:
    def __init__(
        self,
        base_url: str,
        get_jwt: Callable[[], str | None],
        channels: list[str] | None = None,
        max_backoff_seconds: int = 30,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._get_jwt = get_jwt
        self._channels = channels or ["alerts"]
        self._max_backoff = max_backoff_seconds

    def run_forever(self, on_message: Callable[[str], None]) -> None:
        backoff = 0.5
        while True:
            token = self._get_jwt()
            if not token:
                raise RuntimeError("JWT token is required for websocket connection")

            url = f"{self._base_url}?token={token}&channels={','.join(self._channels)}"
            app = WebSocketApp(
                url,
                on_message=lambda _socket, message: on_message(message),
            )
            app.run_forever()
            sleep_for = min(backoff + random.random() * 0.25, self._max_backoff)
            time.sleep(sleep_for)
            backoff = min(backoff * 2, self._max_backoff)

File: aegis-js/src/ws.ts
export type AegisWsOptions = {
  url: string;
  getJwt: () => string | null;
  channels?: string[];
  onMessage: (event: MessageEvent<string>) => void;
  onError?: (error: Event) => void;
  maxBackoffMs?: number;
};

export class AegisWebSocketClient {
  private readonly options: AegisWsOptions;
  private socket: WebSocket | null = null;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private backoffMs = 500;

  constructor(options: AegisWsOptions) {
    this.options = options;
  }

  connect(): void {
    const jwt = this.options.getJwt();
    if (!jwt) {
      throw new Error('JWT is required for websocket connection');
    }

    const channels = (this.options.channels ?? ['alerts']).join(',');
    const url = new URL(this.options.url);
    url.searchParams.set('token', jwt);
    url.searchParams.set('channels', channels);

    this.socket = new WebSocket(url.toString());
    this.socket.onmessage = this.options.onMessage;
    this.socket.onerror = (event) => this.options.onError?.(event);
    this.socket.onopen = () => {
      this.backoffMs = 500;
    };
    this.socket.onclose = () => {
      this.scheduleReconnect();
    };
  }

  disconnect(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    this.socket?.close();
    this.socket = null;
  }

  private scheduleReconnect(): void {
    if (this.reconnectTimer) {
      return;
    }

    const maxBackoff = this.options.maxBackoffMs ?? 30_000;
    const jitter = Math.floor(Math.random() * 250);
    const waitMs = Math.min(this.backoffMs + jitter, maxBackoff);

    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this.connect();
    }, waitMs);

    this.backoffMs = Math.min(this.backoffMs * 2, maxBackoff);
  }
}
